src/ray/raylet/node_manager.cc

```diff
@@ -2020,6 +2019,16 @@ void NodeManager::FinishAssignedActorCreationTask(const ActorID &parent_actor_id
           gcs_client_->Actors().AsyncRegister(actor_notification, update_callback));
     }
   }
+  // Kill newly created actor if parent actor is dead.
+  auto parent_actor_entry = actor_registry_.find(parent_actor_id);
```

Comment:
I think the best place to put this code is actually in the `ALIVE` transition handler in `HandleActorStateTransition`, just to keep all of the state transition code in one place. Unfortunately, this handler is sometimes called locally and sometimes as a callback...to make sure the updates to the actor table are made in the correct order, you could swap all of the calls to `HandleActorStateTransition` and `gcs_client_->Actors().Async...`. So, for instance, this code would now be:
1) `gcs_client_->Actors().Async...` update to make the new actor `ALIVE`.
2) Call `HandleActorStateTransition(...)` locally to handle the `ALIVE` transition.
  a) If the parent is dead, disconnect the new actor. This should trigger a second `gcs_client_->Actors().Async...` update to make the new actor `DEAD`.

I know this is pretty brittle; let us know if you have suggestions on how to improve this.

Reply:
I think simply calling `gcs_client_->Actors().Async...`  is not enough to disconnect the actor. `ProcessDisconnectClientMessage` needs to be called to ensure the tasks for the actor are marked as failed and the actor resources are released. `ProcessDisconnectClientMessage` calls `HandleDisconnectedActor` which in turn calls `HandleActorStateTransition` again. So the code path would be `HandleActorStateTransition -->  ProcessDisconnectClientMessage -->  HandleDisconnectedActor --> HandleActorStateTransition` which is a bit weird. It should work though, I can make that change if it seems fine.

Reply:
Right sorry, I was also suggesting to still call `ProcessDisconnectClientMessage`.

Yeah, it is a bit weird, but I also want to make sure that the updates to the actor table get made in the correct order even if the actor resumed from a checkpoint. The lookup to the checkpoint table before this bit is also asynchronous, and it doesn't add the `ALIVE` entry until the lookup returns, so the actor table updates could get reordered in that case.